commands.py

```diff
@@ -177,111 +176,71 @@ def RGB(r, g, b, colour):
-            s = Fraction(colour[1].replace("%", ""))
-            l = Fraction(colour[2].replace("%", ""))
-
-            if h > 360 or s > 100 or l > 100:
-                return await ctx.send("Please input 0 <= h <= 360, 0 <= s, l <= 100.")
-
-            await ctx.send(embed=hslRGB(h, s, l, content))
-        elif re.fullmatch(r"cmyk\((\d{1,3}(\.\d*)?%?, *){3}\d{1,3}(\.\d*)?%?\)", colour):
-            content = colour
-            colour = colour[5:].split(")")[0].split(",")
-            c = Fraction(colour[0].replace("%", ""))
-            m = Fraction(colour[1].replace("%", ""))
-            y = Fraction(colour[2].replace("%", ""))
-            k = Fraction(colour[3].replace("%", ""))
-
-            if c > 100 or m > 100 or y > 100 or k > 100:
-                return await ctx.send("Please input 0 <= c, m, y, k <= 100.")
-
-            await ctx.send(embed=cmykRGB(c, m, y, k, content))
-        elif colour.lower() in css.keys():
+        elif c_str := re.search(r"#([0-9A-Fa-f]{3}){1,2}", colour):
+            c_str = c_str.group()
+            if len(c_str) == 7:
+                await ctx.send(embed=RGB(int(c_str[1:3], 16), int(c_str[3:5], 16), int(c_str[5:], 16), c_str))
+            elif len(c_str) == 4:
+                await ctx.send(embed=RGB(int(c_str[1] * 2, 16), int(c_str[2] * 2, 16), int(c_str[3] * 2, 16), c_str))
+        elif c_str := re.search(r"([0-9A-Fa-f]{3}){1,2}", colour):
+            c_str = c_str.group()
+            if len(c_str) == 6:
+                await ctx.send(embed=RGB(int(c_str[:2], 16), int(c_str[2:4], 16), int(c_str[4:], 16), c_str))
+            elif len(c_str) == 3:
+                await ctx.send(embed=RGB(int(c_str[0] * 2, 16), int(c_str[1] * 2, 16), int(c_str[2] * 2, 16), c_str))
+        elif c_str := re.search(r"rgb\((\d{1,3}), *(\d{1,3}), *(\d{1,3})\)", colour):
+            r, b, g = map(int, c_str.group(1, 2, 3))
+            if max(r, g, b) > 255 or min(r, g, b) < 0:
+                return await ctx.send("You inputted an invalid colour. Please try again.", delete_after=5)
+            await ctx.send(embed=RGB(r, b, g, c_str))
+        elif c_str := re.search(r"hsl\((\d{1,3}(?:\.\d*)?), *(\d{1,3}(?:\.\d*)?)%?, *(\d{1,3}(?:\.\d*)?)%?\)", colour):
+            h, s, l = map(Fraction, c_str.group(1, 2, 3))
+            if h > 360 or max(s, l) > 100 or min(h, s, l) < 0:
+                return await ctx.send("You inputted an invalid colour. Please try again.", delete_after=5)
+            await ctx.send(embed=hslRGB(h, s, l, c_str))
+        elif c_str := re.search(r"cmyk\((\d{1,3}(\.\d*)?)%?, *(\d{1,3}(\.\d*)?)%?, *(\d{1,3}(\.\d*)?)%?, *(\d{1,3}(\.\d*)?)%?\)", colour):
+            c, m, y, k = map(Fraction, c_str.group(1, 2, 3, 4))
+            if max(c, m, y, k) > 100 or min(c, m, y, k) < 0:
+                return await ctx.send("You inputted an invalid colour. Please try again.", delete_after=5)
+            await ctx.send(embed=cmykRGB(c, m, y, k, c_str))
+        elif colour.lower() in css:
```

Comment:
This now allows the colour specification to be anywhere in the text. Match.group allows for easy selection of the specific values that we are interested in. It does mean that we have to write out the capturing groups more explicitly so that the numbers are saved as we want them. I also improved the way that the valid checks work. I guess that I don't really have to check for <0 since we don't allow a - in the match so perhaps that should be removed.